lib/ip_blocklist.js

```diff
@@ -0,0 +1,165 @@
+    self.filePath = filePath
+    self.fileName = path.basename(filePath)
+
+    var startTime = Date.now()
+    return statFile(filePath)
+      .then(function (fileStats) {
+        self.fileLastModified = fileStats.mtime
+        self.fileSize = fileStats.size
+        return readFile(filePath, 'utf8')
+      })
+      .then(function (data) {
+        return parse(data)
+      })
+      .then(function (rows) {
+        rows.forEach(function (row) {
+          var parsedData = parseRow(row[0])
+
+          if (parsedData) {
+            if (!self.ipsByPrefixLength[parsedData.prefixLength]) {
+              self.prefixLengths.push(parsedData.prefixLength)
+              self.ipsByPrefixLength[parsedData.prefixLength] = {}
+            }
+            self.ipsByPrefixLength[parsedData.prefixLength][parsedData.baseIp] = true
+          }
+        })
+
+        var endTime = Date.now()
+        log.info({
+          op: 'fxa.customs.blocklist.load',
+          fileName: self.fileName,
+          fileSize: self.fileSize,
+          loadedIn: (endTime - startTime)
+        })
+      })
+  }
+
+  IPBlocklist.prototype.clear = function () {
+    this.prefixLengths = []
+    this.ipsByPrefixLength = {}
+  }
+
+  IPBlocklist.prototype.contains = function (ipAddress) {
+    var self = this
+
+    var startTime = Date.now()
+    var endTime
+
+    log.info({
+      op: 'fxa.customs.blocklist.check',
+      ip: ipAddress,
+      fileName: self.fileName,
+      fileSize: self.fileSize
+    })
+
+    try {
+      return self.prefixLengths.some(function (prefixLength) {
+        var result = getBaseIp(ipAddress, prefixLength) in self.ipsByPrefixLength[prefixLength]
+
+        // If `hit` log metrics on time taken
+        if (result) {
+          endTime = Date.now()
+          log.info({op: 'fxa.customs.blocklist.hit', ip: ipAddress, fileName: self.fileName, fileSize: self.fileSize,
+            foundIn: (endTime - startTime)})
+        }
+
+        return result
+      })
+    } catch (err) {
+      // On any error, fail closed and reject request. These errors could be
+      // malformed ip address, etc
+      log.error({op: 'fxa.customs.blocklist.error', fileName: self.fileName, fileSize: self.fileSize, err: err})
+      return true
+    }
+  }
+
+  IPBlocklist.prototype.pollForUpdates = function () {
+    this.stopPolling()
+    this.pollInterval = setInterval(this.refresh.bind(this), config.ipBlocklist.updatePollInterval * 1000)
+  }
+
+  IPBlocklist.prototype.stopPolling = function () {
+    clearInterval(this.pollInterval)
+  }
+
+  IPBlocklist.prototype.refresh = function () {
+    var self = this
+
+    log.trace({op: 'fxa.customs.blocklist.refreshList', fileName: self.fileName, fileSize: self.fileSize})
+
+    return statFile(self.filePath)
+      .then(function (fileStats) {
+        var mtime = fileStats.mtime
+        if (mtime > self.fileLastModified) {
+          self.fileLastModified = mtime
+          return self.load(self.filePath)
```

Comment:
> Reloading on a 1mb file typically takes 300ms on my computer
> [...]
> I took a guess of 10k for batch sizes 

Assuming 16 bytes per line, that's 65k lines in 300ms, or about 45ms per 10k lines.  That's still a long time not to do any I/O.  1k batches maybe?  I dunno, it's all guesswork at this point :-)